rasa/nlu/utils/bilou_utils.py

```diff
@@ -283,10 +311,113 @@ def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
                 for i in range(idx + 1, last_idx):
                     predicted_tags[i] = f"{INSIDE}{tag}"
 
-    return predicted_tags
+    return predicted_tags, predicted_confidences
+
+
+def _tag_to_use(
+    relevant_tags: List[Text], relevant_confidences: List[float]
+) -> Tuple[Text, float]:
+    """Decide what tag to use according to the following metric:
+
+    Calculate the average confidence per tag.
+    Calculate the percentage of tokens assigned to a tag within the entity per tag.
+    The harmonic mean of those two metrics is the score for the tag.
+    The tag with the highest score is taken as the tag for the entity.
+
+    Args:
+        relevant_tags: The tags of the entity.
+        relevant_confidences: The confidence values.
+
+    Returns:
+        The tag to use. The average confidence value of that tag.
+    """
+    # Calculate the average confidence per tag.
+    avg_confidence_per_tag = _avg_confidence_per_tag(
+        relevant_tags, relevant_confidences
+    )
+    # Calculate the percentage of tokens assigned to a tag per tag.
+    token_percentage_per_tag = Counter(relevant_tags)
+    for tag, count in token_percentage_per_tag.items():
+        token_percentage_per_tag[tag] = round(count / len(relevant_tags), 2)
+
+    # Calculate the harmonic mean between the two metrics per tag.
+    score_per_tag = {}
+    for tag, token_percentage in token_percentage_per_tag.items():
+        avg_confidence = avg_confidence_per_tag[tag]
+        score_per_tag[tag] = (
+            2
+            * (avg_confidence * token_percentage)
+            / (avg_confidence + token_percentage)
+        )
+
+    # Take the tag with the highest score as the tag for the entity
+    tag = max(score_per_tag, key=score_per_tag.get)
+    avg_confidence = avg_confidence_per_tag[tag]
+
+    return tag, avg_confidence
```

Comment:
should the score be used as "confidence" then?

Reply:
might make sense, the score would be for sure lower compared to the avg confidence score

Reply:
but that's good I guess, since it is kind of artificial substitution